dataset/batch_image.py

```diff
@@ -1,525 +1,767 @@
-    x, new_x, len_x = calc_coords(image_shape[0], shape[0], crop)
-    y, new_y, len_y = calc_coords(image_shape[1], shape[1], crop)
-    new_x = new_x, new_x + len_x
-    x = x, x + len_x
-    new_y = new_y, new_y + len_y
-    y = y, y + len_y
-
-    return new_x, new_y, x, y
+def transform_actions(prefix='', suffix='', wrapper=None):
+    """ Transforms classmethods that have names like <prefix><name><suffix> to pipeline's actions executed in parallel.
 
+    First, it finds all *class methods* which names have the form <prefix><method_name><suffix>
+    (ignores those that start and end with '__').
 
+    Then, all found classmethods are decorated through ``wrapper`` and resulting
+    methods are added to the class with the names of the form <method_name>.
 
-class BaseImagesBatch(Batch):
-    """ Batch class for 2D images """
-    components = "images", "labels"
+    Parameters
+    ----------
+    prefix : str
+    suffix : str
+    wrapper : str
+        name of the wrapper inside ``Batch`` class
 
-    @property
-    def image_shape(self):
-        """: tuple - shape of the images """
-        return self.images.shape[1:]
+    Examples
+    --------
+    >>> from dataset import ImagesBatch
+    >>> @transform_actions(prefix='_', suffix='_')
+    ... class MyImagesBatch(ImagesBatch):
+    ...     @classmethod
+    ...     def _flip_(cls, image):
+    ...             return image[:,::-1]
 
-    def assemble(self, all_res, *args, **kwargs):
-        """ Assemble the batch after a parallel action """
-        _ = all_res, args, kwargs
-        raise NotImplementedError("Use ImagesBatch")
+    Note that if you only want to redefine actions you still have to decorate your class.
 
-    def _assemble_load(self, all_res, *args, **kwargs):
-        """ Build the batch data after loading data from files """
-        _ = all_res, args, kwargs
-        return self
+    >>> from dataset.opensets import CIFAR10
+    >>> dataset = CIFAR10(batch_class=MyImagesBatch, path='.')
 
-    @action
-    def load(self, src, fmt=None, components=None, *args, **kwargs):
-        """ Load data """
-        return super().load(src, fmt, components, *args, **kwargs)
+    Now dataset.pipeline has flip action that operates as described above.
+    If you want to apply an action with some probability, then specify ``p`` parameter:
 
-    @action
-    def dump(self, dst, fmt=None, components=None, *args, **kwargs):
-        """ Save data to a file or a memory object """
-        return super().dump(dst, fmt, components=None, *args, **kwargs)
+    >>> from dataset import Pipeline
+    >>> pipeline = (Pipeline()
+    ...                 ...preprocessing...
+    ...                 .flip(p=0.7)
+    ...                 ...postprocessing...
 
-    @action
-    def noop(self):
-        """ Do nothing """
-        return self
+    Now each image will be flipped with probability 0.7.
+    """
+    def __decorator(cls):
```

Comment:
Why two underscores? One is just enough.

Reply:
+